add karmada-v1.15 blog (#887)
karmada-io/website: blog/2025-09-05-karmada-v1.15/karmada-v1.15.md

# Karmada v1.15 Released! Enhanced Resource Awareness for Multi-Template Workloads!

Karmada is an open multi-cloud and multi-cluster container orchestration engine designed to help users deploy and operate business applications in a multi-cloud environment. With its compatibility with the native Kubernetes API, Karmada can smoothly migrate single-cluster workloads while still maintaining coordination with the surrounding Kubernetes ecosystem tools.

[Karmada v1.15](https://github.com/karmada-io/karmada/releases/tag/v1.15.0) has been released, this version includes the following new features:

- Precise resource awareness for multi-template workloads
- Enhanced cluster-level failover functionality
- Structured logging
- Significant performance improvements for Karmada controllers and schedulers

## Overview of New Features

### Precise Resource Awareness for Multi-Template Workloads

> Karmada utilizes a resource interpreter to retrieve the replica count and resource requests of workloads. Based on this data, it calculates the total resource requirements of the workloads, thereby enabling advanced capabilities such as resource-aware scheduling and federated quota management. This mechanism works well for traditional single-template workloads. However, workload Custom Resource Definitions (CRDs) for many AI and big data applications (e.g., FlinkDeployments, PyTorchJobs, and RayJobs) consist of multiple Pod templates or components, each with unique resource demands. Since the resource interpreter can only process resource requests from a single template and fails to accurately reflect differences between multiple templates, the resource calculation for multi-template workloads is not precise enough.

Reviewer: ```suggestion
Karmada utilizes a resource interpreter to retrieve the replica count and resource requests of workloads. Based on this data, it calculates the total resource requirements of the workloads, thereby enabling advanced capabilities such as resource-aware scheduling and federated quota management. This mechanism works well for traditional single-template workloads. However, workloads for many AI and big data applications (e.g., FlinkDeployments, PyTorchJobs, and RayJobs) consist of multiple Pod templates or components, each with unique resource demands. Since the resource interpreter can only process resource requests from a single template and fails to accurately reflect differences between multiple templates, the resource calculation for multi-template workloads is not precise enough.
```

Karmada v1.15 introduces support for structured logging, which can be configured to output in JSON format using the `--logging-format=json` startup parameter. An example of structured logging is as follows:

```json
{
  "ts":"log timestamp",
  "logger":"cluster_status_controller",
  "level": "info",
  "msg":"Syncing cluster status",
  "clusterName":"member1"
}
```

The introduction of structured logging significantly improves the usability and observability of logs:

- **Efficient Integration**: Seamless integration with mainstream logging systems such as Elastic, Loki, and Splunk, without relying on complex regular expressions or log parsers.

- **Efficient Query**: Structured fields support fast retrieval and analysis, significantly improving troubleshooting efficiency.

- **Enhanced Observability**: Key context information (e.g., cluster name, log level) is presented as structured fields, facilitating cross-component and cross-time event correlation for accurate issue localization.

> - **Improved Maintainability**: Structured logging makes it easier for developers and operations personnel to maintain, parse, and adjust log formats during system evolution, ensuring long-term stability and consistency of the logging system.

Reviewer: ```suggestion
- **Maintainability**: Structured logging makes it easier for developers and operators to maintain, parse, and evolve log formats as the system changes.
```

### Significant Performance Improvements for Karmada Controllers and Schedulers

In this version, the Karmada performance optimization team has continued to focus on improving the performance of Karmada's key components, achieving significant progress in both controllers and schedulers.

> **In terms of the controller**, by introducing a priority queue, the controller can give priority to responding to user-triggered resource changes after a restart or active/standby switchover, thereby significantly reducing the downtime during service restart and failover processes.

Reviewer: ```suggestion
**In terms of the controller**, by introducing [controller-runtime priority queue](https://github.com/kubernetes-sigs/controller-runtime/issues/2374), the controller can give priority to responding to user-triggered resource changes after a restart or leader transition, thereby significantly reducing the downtime during service restart and failover processes.
```

> The test environment included 5,000 Deployments, 2,500 Policies, and 5,000 ResourceBindings. With the controller restarted and a large number of pending events remaining in the work queue, Deployments and Policies were updated. Test results showed that **the controller could immediately respond to and prioritize processing these update events, verifying the effectiveness of this optimization**.

Reviewer: ```suggestion
The test environment included 5,000 Deployments, 2,500 Policies, and 5,000 ResourceBindings. The Deployment and Policy were updated when the controller restarted with a large number of pending events still in the work queue. Test results showed that **the controller could immediately respond to and prioritize processing these update events, verifying the effectiveness of this optimization**.
```
